### PR TITLE
Media: Implement media uploads

### PR DESCRIPTION
**Pull request by kean**
Part of #21457

## To test:

This is not yet the final version of the feature, but it covers most of the basic media upload functionality, including uploading photos and videos and other documents from all available sources.

Some of the tests from https://github.com/wordpress-mobile/WordPress-iOS/pull/21343 and https://github.com/wordpress-mobile/WordPress-iOS/pull/21411 can be re-run.

https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/83b3eeb1-99d0-46e0-913c-db274811c257

https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/fdd047f4-8166-4c8f-9270-e112f768e59a

## Thumbnails

In the [original PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/21470), I re-implemented `MediaThumbnailService` using Async/Await with only minor changes to the logic (e.g. adding cancellation and decompression support). But, there turned out to be many other inefficiencies in the existing logic.

So, the way it currently works is as follows:

- For local assets that were just imported to the library, it waits until `MediaImportService` generates the `absoluteThumbnailLocalURL`. Then the cell refreshes.
- For remote assets, it just fetches the remote image, stores it locally, and sets `absoluteThumbnailLocalURL`

So the source of thumbnails is now always the same: `absoluteThumbnailLocalURL`.

I also made a number of other performance improvements: removed `URLCache.shared` usage, removed any unnecessary delays, removed the code that was transcoding the thumbnails, losing quality and performing duplicated work, etc. It should now be as fast as it gets as it can be at serving the thumbnails.

## Regression Notes
1. Potential unintended areas of impact: n/a (doesn't affect the production code)
2. What I did to test those areas of impact (or what existing automated tests I relied on): manual testing
3. What automated tests I added (or what prevented me from doing so): n/a

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Media/Library/MediaCollectionCell.swift [hunk 1]
```diff
@@ -1,20 +1,30 @@
...
     private let imageView = UIImageView()
+    private let overlayView = CircularProgressView()
     private var viewModel: MediaCollectionCellViewModel?
+    private var cancellables: [AnyCancellable] = []
 
     override init(frame: CGRect) {
         super.init(frame: frame)
 
-        backgroundColor = .systemGroupedBackground
+        backgroundColor = .secondarySystemBackground
```
Comment (kean) [inline WordPress/Classes/ViewRelated/Media/Library/MediaCollectionCell.swift:13]: `systemGroupedBackground` was not working as expected in dark mode.

File: WordPress/Classes/ViewRelated/Media/Library/MediaCollectionCellViewModel.swift [hunk 2]
```diff
@@ -24,51 +27,112 @@ final class MediaCollectionCellViewModel {
...
+        isPrefetchingNeeded = false
+        cancelThumbnailRequestIfNeeded()
+    }
+
+    // MARK: - Thumbnail
 
-        requestCount += 1
-        guard requestCount == 1 else {
+    private func fetchThumbnailIfNeeded() {
+        guard isVisible || isPrefetchingNeeded else {
```
Comment (kean) [inline WordPress/Classes/ViewRelated/Media/Library/MediaCollectionCellViewModel.swift:72]: This is better than relying on `requestCount`.

File: WordPress/Classes/ViewRelated/Media/Library/MediaCollectionCellViewModel.swift [hunk 3]
```diff
@@ -24,51 +27,112 @@ final class MediaCollectionCellViewModel {
...
+    // Monitors thumbnails generated by `MediaImportService`.
+    private func didUpdateLocalThumbnail() {
+        guard media.remoteStatus != .sync, media.localThumbnailURL != nil else { return }
+        fetchThumbnailIfNeeded()
+    }
+
+    // MARK: - Status
+
+    private func updateOverlayState() {
+        switch media.remoteStatus {
```
Comment (kean) [inline WordPress/Classes/ViewRelated/Media/Library/MediaCollectionCellViewModel.swift:127]: It now tracks upload progress directly without even needing `MediaCoordinator`, which allowed me to significantly simplify the code compared to the previous versions.

**Review by momo-ozawa — APPROVED** [review]
I followed testing instructions from previous PRs. Works as described!

**Pull request merged**